fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+            clip_segments,
+            download_dir_list,
+            ext_list,
+            buffer_seconds_list,
+        )
+    )
+
+
+def _parse_list_arg(arg, list_len):
+    if isinstance(arg, list):
+        if len(arg) != list_len:
+            # If arg list is not the right length, ignore it and set all arg to
+            # `None`
+            arg_list = [None] * list_len
+        else:
+            arg_list = arg
+    else:
+        # If a non-list arg is given, repeat it for the length of the list
+        arg_list = [arg] * list_len
+
+    return arg_list
+
+
+def _single_thread_download(tasks, max_videos):
+    downloaded = []
+    errors = defaultdict(list)
+    with fou.ProgressBar(total=max_videos, iters_str="videos") as pb:
+        for task in tasks:
+            is_success, url, error_type = _do_download(task)
+            if is_success:
+                downloaded.append(url)
+                pb.update()
+                if len(downloaded) >= max_videos:
+                    return downloaded, errors
+            else:
+                errors[error_type].append(url)
+    return downloaded, errors
+
+
+def _multi_thread_download(tasks, max_videos, num_workers):
+    downloaded = []
+    errors = defaultdict(list)
+    with fou.ProgressBar(total=max_videos, iters_str="videos") as pb:
+        with multiprocessing.dummy.Pool(num_workers) as pool:
+            for is_success, url, error_type in pool.imap_unordered(
+                _do_download, tasks
+            ):
+                if is_success:
+                    if len(downloaded) < max_videos:
+                        downloaded.append(url)
+                        pb.update()
+                    else:
+                        return downloaded, errors
+                else:
+                    errors[error_type].append(url)
+    return downloaded, errors
+
+
+def _build_ydl_opts(ext, video_path, clip_segment, download_dir):
+    ytdl_logger = logging.getLogger("ytdl-ignore")
+    ytdl_logger.disabled = True
+    ydl_opts = {
+        "age_limit": 99,
+        "logger": ytdl_logger,
+    }
+
+    if video_path:
+        output_path = video_path
+        _ext = os.path.splitext(os.path.basename(video_path))[1]
+        _ext = _ext.lstrip(".")
+
+    else:
+        if ext is not None:
+            _ext = ext.lstrip(".")
+            path_ext = _ext
+        else:
+            _ext = None
+            path_ext = "%(ext)s"
+        path_name = "%(title)s"
+        output_path = os.path.join(
+            download_dir, "%s.%s" % (path_name, path_ext)
+        )
+
+    if _ext:
+        ydl_opts["format"] = _ext
+
+    ydl_opts["outtmpl"] = output_path
+
+    return ydl_opts, output_path
+
+
+def _do_download(args):
+    url, video_path, clip_segment, download_dir, ext, buffer_seconds = args
+    ydl_opts, output_path = _build_ydl_opts(
+        ext, video_path, clip_segment, download_dir
+    )
+
+    try:
+        if clip_segment in [None, [None, None], [0, None]]:
```

Comment:
Need to be careful that `[None, None] != (None, None)`. I'd recommend doing that parsing in `_build_tasks_list()` so that, by the time we get here, `clip_segment` is either `None` or `(start, stop)`.

Reply:
^example of task creation logic bleeding into the download implementation. Standardize as much as possible in `_build_tasks_list()` for simplicitly.